### PR TITLE
Run function that writes the XML to a file and sends nmap stdout to the console

### PR DESCRIPTION
**Pull request by NoF0rte**
With the project I am working on, I wanted to be able to basically run nmap with its normal console output while being able to get the parsed `Run` struct.

With the current code. this could be achieved by using the `RunWithStreamer` function but there were two problems with that for me:
1. This function creates a pipe for the nmap command's stdout so that it can be "piped" into the `Streamer` instance that was passed in.  Since I want to simply output to `os.Stdout`, I would have to pass in a `Streamer` instance that just writes back to `os.Stdout` which is convoluted.
2. I didn't like that "--stats-every 5s" was added to the nmap args.

I added the `RunDirect` function to meet my needs. It sets the nmap command's stdout to `os.Stdout`, adds "-oX file" to the nmap arguments (like `RunWithStreamer` does) and then reads and parses the XML output file.

**Note:** I was having a hard time thinking of a good function name so feel free to change it.

### CODE REVIEW

**Comment by elivlo** [comment]
I understand the problems the method `RunWithStreamer` causes. I like the changes :+1: 
But I have some notes for code quality :)

- The two new functions should have some unit tests
- Please give an hint on how to run the method with a simple script in the `example` directory

File: nmap.go [hunk 1]
```diff
@@ -394,6 +316,100 @@ func (s *Scanner) RunAsync() error {
...
+			return nil, warnings, err
+		}
+
+		return s.processOutput(xml, stderr.Bytes())
+	}
+}
+
+func (s *Scanner) processOutput(xml []byte, stderr []byte) (result *Run, warnings []string, err error) {
+	// Process nmap stderr output containing none-critical errors and warnings
+	// Everyone needs to check whether one or some of these warnings is a hard issue in their use case
```
Comment (elivlo) [inline nmap.go:372]: Please end this sentences with a dot
```suggestion
	// Process nmap stderr output containing none-critical errors and warnings.
	// Everyone needs to check whether one or some of these warnings is a hard issue in their use case.
```

File: nmap.go [hunk 2]
```diff
@@ -394,6 +316,100 @@ func (s *Scanner) RunAsync() error {
...
+
+	// Call filters if they are set.
+	if s.portFilter != nil {
+		result = choosePorts(result, s.portFilter)
+	}
+	if s.hostFilter != nil {
+		result = chooseHosts(result, s.hostFilter)
+	}
+
+	// Return result, optional warnings but no error
```
Comment (elivlo) [inline nmap.go:409]:
```suggestion
	// Return result, optional warnings but no error.
```

File: nmap.go [hunk 3]
```diff
@@ -394,6 +316,100 @@ func (s *Scanner) RunAsync() error {
...
+// RunDirect runs nmap synchronously. The XML output is written directly to a file and stdout is written to the console.
+func (s *Scanner) RunDirect(file string) (result *Run, warnings []string, err error) {
+	var stderr bytes.Buffer
+
+	args := s.args
+
+	// Enable XML output
+	args = append(args, "-oX")
+
+	// Get XML output in stdout instead of writing it in a file
```
Comment (elivlo) [inline nmap.go:328]:
```suggestion
	// Get XML output in stdout instead of writing it in a file.
```

File: nmap.go [hunk 4]
```diff
@@ -394,6 +316,100 @@ func (s *Scanner) RunAsync() error {
...
+		return nil, warnings, err
+	}
+
+	// Make a goroutine to notify the select when the scan is done.
+	done := make(chan error, 1)
+	go func() {
+		done <- cmd.Wait()
+	}()
+
+	// Wait for nmap process or timeout
```
Comment (elivlo) [inline nmap.go:351]:
```suggestion
	// Wait for nmap process or timeout.
```

**Review by elivlo — CHANGES_REQUESTED** [review]

**Review by Ullaakut — COMMENTED** [review]
I understand the problem this solves, and also like the changes, but I'm feeling like the library is getting more and more cluttered. I'm not sure what to do about this, maybe the problem is a fundamental one, and the library should have been initially designed around being able to run nmap synchronously or asynchronously in a more seamless way.

This makes me think it might be time for a redesign/rewrite soon. In the meantime, I'm fine with this PR getting merged, though I think it calls for a deeper change in the long term.

What do you think @elivlo ?

**Comment by elivlo** [comment]
@Ullaakut I agree with you! That is a similar intention I had in the previous PR and didn't want to merge it.

I think that the redesign of this library could take some time. We need to make sure that all the current functionality will be implemented while reducing the amount of `Run` functions.

I am finishing my Bachelor in a few weeks, so I need some time before I will start. Maybe we could open a Project/EPIC meanwhile and gather some ideas?

**Comment by elivlo** [comment]
> I'm going to start a new job in the next few days/weeks on my side so the timing isn't great :( Congrats on your bachelor man!
> 
> If you'd like, I'll put a reminder in my calendar to get back to this when I have more time. How does that sound?

@Ullaakut 
Thanks! I have opened the PR #108 for this purpose and made many breaking changes there. But before I will change the documentation, comments and release it, I would like to get some feedback.

**Comment by elivlo** [comment]
Hey @NoF0rte :)

your request should be solved now that we did the Major version change.

**Comment by NoF0rte** [comment]
Technically my original issue is now resolved with v3 but a new one arises with this
```go
// Write XML to standard output.
// If toFile is set then write XML to file and normal nmap output to stdout.
if s.toFile != nil {
  args = append(args, "-oX", *s.toFile, "-oN", "-")
} else {
  args = append(args, "-oX", "-")
}
```

May I ask why it now forces the normal output to stdout like that? Doesn't nmap always send the normal output to stdout if you don't specify another output option to go to stdout?

**Comment by elivlo** [comment]
You're right, it is not necessary to force it. And since you can call `WithNmapOutput` it could cause problems.
You may open a PR for that, and I will review and merge it quickly.

**Comment by NoF0rte** [comment]
Just created #110. Thanks for the quick responses by the way. I appreciate it.

**Comment by NoF0rte** [comment]
You can probably close this PR too

**Comment by elivlo** [comment]
> Just created #110. Thanks for the quick responses by the way. I appreciate it.

Thanks! I always try to respond fast, but sometimes I miss the email in my mailbox :)

Alright! I close it now. The issue will be resolved here: #110

**Pull request closed without merging**